packages/@blockprotocol/hook/src/react.ts

```diff
@@ -0,0 +1,246 @@
+    ConstructorParameters<typeof HookBlockHandler>[0],
+    "element"
+  >,
+): { hookService: HookBlockHandler | null } => {
+  return useHookServiceConstructor({
+    Handler: HookBlockHandler,
+    constructorArgs,
+    ref,
+  });
+};
+
+/**
+ * Create a HookBlockHandler instance, using a reference to an element in the
+ * block.
+ *
+ * The hookService will only be reconstructed if the element reference changes.
+ * Updates to any callbacks after first constructing should be made by
+ * calling hookService.on("messageName", callback);
+ */
+export const useHookEmbedderService = (
+  ref: RefObject<HTMLElement>,
+  constructorArgs?: Omit<
+    ConstructorParameters<typeof HookEmbedderHandler>[0],
+    "element"
+  >,
+): { hookService: HookEmbedderHandler | null } => {
+  return useHookServiceConstructor({
+    Handler: HookEmbedderHandler,
+    ref,
+    constructorArgs,
+  });
+};
+
+type Hook<T extends HTMLElement> = {
+  id: string | null;
+  teardown: (() => Promise<void>) | null;
+  params: {
+    service: HookBlockHandler | null;
+    node: T;
+    type: string;
+    path: string;
+  };
+};
+
+export const useHook = <T extends HTMLElement>(
+  service: HookBlockHandler | null,
+  ref: RefObject<T | null | void>,
+  type: string,
+  path: string,
+  fallback: (node: T) => void | (() => void),
+) => {
+  const hookRef = useRef<null | Hook<T>>(null);
+  const [, setError] = useState();
+
+  const fallbackRef = useRef(fallback);
+
+  useLayoutEffect(() => {
+    fallbackRef.current = fallback;
+  });
+
+  useLayoutEffect(() => {
+    return () => {
+      hookRef.current?.teardown?.().catch((err) => {
+        setError(() => {
+          throw err;
+        });
+      });
+    };
+  }, []);
+
+  useLayoutEffect(() => {
+    const existingHook = hookRef.current?.params;
+    const node = ref.current;
+
+    if (
+      existingHook &&
+      existingHook.service === service &&
+      existingHook.node === node &&
+      existingHook.path === path &&
+      existingHook.type === type
+    ) {
+      return;
+    }
+
+    const teardownPromise =
+      hookRef.current?.teardown?.().catch() ?? Promise.resolve();
+
+    if (node && service) {
+      const controller = new AbortController();
+
+      const reuseId =
+        existingHook &&
+        existingHook.service === service &&
+        existingHook.path === path &&
+        existingHook.type === type;
+
+      const hook: Hook<T> = {
+        id: reuseId ? hookRef.current?.id ?? null : null,
+        params: {
+          service,
+          type,
+          path,
+          node,
+        },
+        teardown: async () => {
+          controller.abort();
+
+          if (hook.id) {
+            try {
+              hook.id = null;
+              if (hookRef.current === hook) {
+                hookRef.current = null;
+              }
+
+              if (!service.destroyed) {
+                await service.hook({
+                  data: {
+                    hookId: hook.id,
```

Comment:
This was set `null` a few lines above

Reply:
Good catch https://github.com/blockprotocol/blockprotocol/commit/567ed9c2fe9908a5f5447019c1bc3480f2c57258